crates/dbs-virtio-devices/src/block/ufile/localfile.rs

```diff
@@ -132,3 +132,344 @@ impl<E: IoEngine + Send> Ufile for LocalFile<E> {
+            assert_eq!(element.1, VIRTIO_BLK_S_OK);
+        }
+
+        // Test with IoUring.
+        let file = TempFile::new().unwrap();
+        let iouring = IoUring::new(file.as_file().as_raw_fd(), 128).unwrap();
+        let mut helper = EpollHelper::new().unwrap();
+        helper
+            .add_event(iouring.event_fd().as_raw_fd(), 0xfeed)
+            .unwrap();
+        let mut file_with_iouring = LocalFile::new(file.into_file(), false, iouring).unwrap();
+        let mut handler = TestHandler;
+        let buf = vec![0xffu8; 0xff];
+        file_with_iouring
+            .io_write_submit(
+                8,
+                &mut vec![IoDataDesc {
+                    data_addr: buf.as_ptr() as u64,
+                    data_len: 0x8_usize,
+                }],
+                0x12,
+            )
+            .unwrap();
+        helper.run(&mut handler).unwrap();
+        let res = file_with_iouring.io_complete().unwrap();
+
+        for element in res {
+            assert_eq!(element.0, 0x12);
+            assert_eq!(element.1, VIRTIO_BLK_S_OK);
+        }
+    }
+
+    #[test]
+    fn test_io_read_submit() {
+        // Test with Aio.
+        let file = TempFile::new().unwrap();
+        file.as_file().seek(SeekFrom::Start(0x120)).unwrap();
+        file.as_file().write_all(&[0xeeu8; 0x20]).unwrap();
+        let aio = Aio::new(file.as_file().as_raw_fd(), 128).unwrap();
+        let mut file_with_aio = LocalFile::new(file.into_file(), false, aio).unwrap();
+        let rbuf = vec![0u8; 0x100];
+        let ret = file_with_aio.io_read_submit(
+            -0x20,
+            &mut vec![IoDataDesc {
+                data_addr: rbuf.as_ptr() as u64,
+                data_len: 0x20,
+            }],
+            0x456,
+        );
+        assert_eq!(ret.unwrap_err().kind(), io::ErrorKind::InvalidInput);
+
+        file_with_aio
+            .io_read_submit(
+                0x120,
+                &mut vec![IoDataDesc {
+                    data_addr: rbuf.as_ptr() as u64,
+                    data_len: 0x20,
+                }],
+                0x456,
+            )
+            .unwrap();
+        let com_res = file_with_aio.io_complete().unwrap();
+        for element in com_res {
+            assert_eq!(element.0, 0x456);
+            assert_eq!(element.1, VIRTIO_BLK_S_OK);
+        }
+        assert_eq!(&rbuf[..0x20], &vec![0xee; 0x20]);
+
+        // Test with IoUring.
+        let file = TempFile::new().unwrap();
+        file.as_file().seek(SeekFrom::Start(0x120)).unwrap();
+        file.as_file().write_all(&[0xeeu8; 0x20]).unwrap();
+        let iouring = IoUring::new(file.as_file().as_raw_fd(), 128).unwrap();
+        let mut helper = EpollHelper::new().unwrap();
+        helper
+            .add_event(iouring.event_fd().as_raw_fd(), 0xfeed)
+            .unwrap();
+        let mut file_with_iouring = LocalFile::new(file.into_file(), false, iouring).unwrap();
+        let mut handler = TestHandler;
```

Comment:
We really need to use a handler and epoll helper here?

Reply:
Otherwise the operations would fail. This method of testing is referenced from the UTs wrote in IoUring.